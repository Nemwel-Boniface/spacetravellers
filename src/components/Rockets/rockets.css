/* Will contain styling for the rockets component */
