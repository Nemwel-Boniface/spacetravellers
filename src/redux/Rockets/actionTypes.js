export const FETCHROCKETS = 'FETCHROCKETS';
export const BOOKROCKET = 'bookRocket';
export const CANCELBOOKING = 'CancelBooking';
