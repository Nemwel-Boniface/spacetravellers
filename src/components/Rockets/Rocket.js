import React from "react";
import PropTypes from 'prop-types';
import './rockets.css';

const Rocket = (props) => {
  const { rocket } = props;
  const { id, name, description, image } = rocket;
  return (
    <div id={id} className="rocket">
      <div className="rocketLeft">
        <img src={image} alt="rocket"></img>
      </div>
      <div className="rocketRight">
        <h2>{name}</h2>
        <p>{description}</p>
        <button id={id} type="button">Reserve Rocket</button>
      </div>
    </div>
  )
}
export default Rocket;