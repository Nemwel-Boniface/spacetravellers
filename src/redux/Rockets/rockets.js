import * as actions from './actionTypes';

const baseURL = 'https://api.spacexdata.com/v3/rockets';

const initialRockets = [];

const RocketsReducer = (state = initialRockets, action) => {
  switch (action.type) {
    case actions.FETCHROCKETS:
      return [
        ...action.payLoad,
      ];
    case actions.BOOKROCKET:
      return [...state.map(rocket => {
        if(rocket.id !== action.payLoad)
            return rocket;
        return { ...rocket, reserved: true };
      })
    ]
    case actions.CANCELBOOKING:
      return [...state.map(rocket => {
        if(rocket.id !== action.payLoad)
            return rocket;
        return { ...rocket, reserved: false };
        })
      ]
    default:
      return state;
  }
};

export const getRocketFromAPI = () => (dispatch) => fetch(baseURL)
  .then((res) => res.json()).then((data) => {
    const rockets = data.map((rocket) => ({
      id: rocket.id,
      name: rocket.rocket_name,
      description: rocket.description,
      image: rocket.flickr_images[0],
      reserved: false,
    }));
    dispatch({ type: actions.FETCHROCKETS, payLoad: rockets });
  }).catch(() => {});


export const RocketBooking = (id) => {
  return {
    type: actions.BOOKROCKET,
    payLoad: id,
  }
}

export const CancelRocketBooking = (id) => {
  return {
    type: actions.CANCELBOOKING,
    payLoad: id,
  }
}


export default RocketsReducer;
