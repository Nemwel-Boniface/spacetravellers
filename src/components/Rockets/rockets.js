import React, { useEffect } from 'react';
import Proptypes from 'prop-types';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { getRocketFromAPI } from '../../redux/Rockets/rockets';
import Rocket from './Rocket';

const Rockets = () => {
  const rockets = useSelector((state) => state.RocketsReducer, shallowEqual);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getRocketFromAPI());
  }, []);
  return (
    <div className="rockets">
      <ul>
        {rockets.map((rocket) => (
          <Rocket
            key={rocket.id}
            rocket={rocket}
            image={rocket.image}
            description={rocket.description}
            name={rocket.name}
            reserved={rocket.reserved}
          />
        ))}
      </ul>
    </div>
  );
};
Rockets.propTypes = {
  rockets: Proptypes.arrayOf(
    Proptypes.shape({}),
  ).isRequired,
};
export default Rockets;
