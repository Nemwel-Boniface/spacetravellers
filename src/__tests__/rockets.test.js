import React from "react";
import { Provider } from "react-redux";
import renderer from 'react-test-renderer';
import store from '../redux/configStore';
import Rockets from '../components/Rockets/rockets';
import Rocket from "../components/Rockets/Rocket";

describe('A tests for my Rockets component', () => {
  it('Check whether Rockets component renders', () => {
    const rocket = {
      id: 1,
      image: 'https://img.icons8.com/external-xnimrodx-blue-xnimrodx/64/000000/external-rocket-online-marketing-xnimrodx-blue-xnimrodx.png',
      name: 'falcon1',
      description: 'This is a description about Falcon 1',
      reserved: false,
    };
    const rocketTree = renderer.create(
      <Provider store={store}>
        <Rocket
          key={rocket.id}
          rocket={rocket}
          image={rocket.image}
          description={rocket.description}
          name={rocket.name}
          reserved={rocket.reserved}
        />
      </Provider>
    );
    expect(rocketTree).toMatchSnapshot();
  });


})