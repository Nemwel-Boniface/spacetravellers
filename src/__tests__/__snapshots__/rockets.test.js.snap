// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A tests for my Rockets component Check whether Rockets component renders 1`] = `
<div
  className="rocket"
  id={1}
>
  <div
    className="rocketLeft"
  >
    <img
      alt="rocket"
      src="https://img.icons8.com/external-xnimrodx-blue-xnimrodx/64/000000/external-rocket-online-marketing-xnimrodx-blue-xnimrodx.png"
    />
  </div>
  <div
    className="rocketRight"
  >
    <h2>
      falcon1
    </h2>
    <p>
      This is a description about Falcon 1
    </p>
    <button
      id={1}
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;
